File: packages/core/src/markers/cli/precommit.ts
// Precommit orchestrator + PR-summary formatter (spec 10.1, 10.2; Phase 8).
//
// These are PURE functions: they take the already-computed results of the
// `validate-ledger` (staged) and `scan --policy-mode <mode>` command cores and
// decide BLOCK vs WARN vs OK (10.1), or render a human-readable PR summary from a
// freshness status object (10.2). They never shell out, never touch the
// filesystem, and never re-run the scan/ledger logic -- the thin precommit shell
// script and the CI workflow run the CLI commands and feed their results here.
//
// Block vs warn (spec 10.1):
//   BLOCK  validate-ledger failed (non-append edit, bad/unsigned/invalid proof,
//          registry conflict, schema failure), OR scan found an INVALID row
//          (malformed marker, duplicate slug, unclosed/mismatched end,
//          unsupported inferred marker, unregistered current marker), OR scan's
//          ledger/registry validation failed, OR any row is freshness
//          policy-blocked (release: a required row that is not FRESH).
//   WARN   a SUSPECT / UNPROVEN / UNBOUND row that the policy does NOT block --
//          printed loudly with its required action, but the commit proceeds.
//   OK     nothing of concern.
import type { FreshnessRowOut, FreshnessStatus } from "../freshness.js";
import type { ScanCommandResult } from "./scan.js";
import type { ValidateLedgerCommandResult } from "./validateLedger.js";

export type PrecommitDecision = "BLOCK" | "WARN" | "OK";

export interface PrecommitBlockReason {
  source: "validate-ledger" | "scan";
  code: string;
  row_id?: string;
  message: string;
}

export interface PrecommitWarning {
  row_id: string;
  status: "SUSPECT" | "UNPROVEN" | "UNBOUND";
  reason: string;
  required_action: string;
  // The loud, multi-line warning block (spec 10.1).
  message: string;
}

export interface PrecommitResult {
  decision: PrecommitDecision;
  // 1 when BLOCK (so the git hook aborts), 0 for WARN/OK (commit proceeds).
  exit_code: number;
  block_reasons: PrecommitBlockReason[];
  warnings: PrecommitWarning[];
  // All human-readable lines (block reasons followed by loud warnings), for the
  // hook/CI to print directly.
  messages: string[];
}

// Only the fields the orchestrator reads from each command result. Accepting the
// full result types is fine too -- these slices keep the contract explicit and
// the unit tests hand-buildable.
export interface PrecommitInput {
  validateLedger: Pick<
    ValidateLedgerCommandResult,
    "ok" | "exit_code" | "evidence_valid" | "registry_valid" | "append_only" | "errors"
  >;
  scan: Pick<ScanCommandResult, "exit_code" | "registry_valid" | "evidence_valid" | "status">;
}

const WARN_STATUSES: ReadonlySet<string> = new Set(["SUSPECT", "UNPROVEN", "UNBOUND"]);

function firstReasonCode(row: FreshnessRowOut): string {
  return row.reasons[0]?.code ?? row.status;
}

function defaultRequiredAction(row: FreshnessRowOut): string {
  if (row.required_action) {
    return row.required_action;
  }
  return row.status === "UNBOUND"
    ? `ucp bind --row ${row.row_id}`
    : `ucp prove --row ${row.row_id}`;
}

// Build the loud warning block exactly as spec 10.1 mandates for SUSPECT (the
// status word and reason vary by row).
export function formatPrecommitWarning(row: FreshnessRowOut): PrecommitWarning {
  const reason = firstReasonCode(row);
  const requiredAction = defaultRequiredAction(row);
  const message = [
    `USE-CASE ROW ${row.status}`,
    `row: ${row.row_id}`,
    `reason: ${reason}`,
    `required action: ${requiredAction}`
  ].join("\n");
  return {
    row_id: row.row_id,
    status: row.status as PrecommitWarning["status"],
    reason,
    required_action: requiredAction,
    message
  };
}

export function decidePrecommit(input: PrecommitInput): PrecommitResult {
  const blockReasons: PrecommitBlockReason[] = [];
  const warnings: PrecommitWarning[] = [];

  // --- validate-ledger: any failure blocks (spec 10.1 items 6-10). ---
  if (!input.validateLedger.ok) {
    if (input.validateLedger.errors.length > 0) {
      for (const error of input.validateLedger.errors) {
        blockReasons.push({
          source: "validate-ledger",
          code: error.code,
          message: `validate-ledger (${error.scope}): ${error.message}`
        });
      }
    } else {
      blockReasons.push({
        source: "validate-ledger",
        code: "LEDGER_INVALID",
        message: "validate-ledger reported the evidence ledger or binding registry as invalid"
      });
    }
  }

  // --- scan: ledger/registry validation failures (exit 4) block too. ---
  if (!input.scan.registry_valid || !input.scan.evidence_valid) {
    const ledgerIntegrity = input.scan.status.integrity_errors.filter(
      (error) => error.row_id === undefined
    );
    if (ledgerIntegrity.length > 0) {
      for (const error of ledgerIntegrity) {
        blockReasons.push({
          source: "scan",
          code: error.code,
          message: `scan ledger/registry: ${error.message ?? error.code}`
        });
      }
    } else {
      blockReasons.push({
        source: "scan",
        code: "LEDGER_INVALID",
        message: "scan reported the evidence ledger or binding registry as invalid"
      });
    }
  }

  // --- scan: usage/config/internal error (exit 2) blocks. ---
  if (input.scan.exit_code === 2) {
    blockReasons.push({
      source: "scan",
      code: "SCAN_USAGE_ERROR",
      message: "scan failed with a usage/config/internal error (exit 2)"
    });
  }

  // --- per-row: INVALID blocks; policy-blocked rows block; warn statuses warn. ---
  for (const row of input.scan.status.rows) {
    if (row.status === "INVALID") {
      blockReasons.push({
        source: "scan",
        code: firstReasonCode(row),
        row_id: row.row_id,
        message: `INVALID row ${row.row_id}: ${firstReasonCode(row)}`
      });
      continue;
    }
    if (row.policy_block) {
      // A non-INVALID row the policy blocks: in release mode, a required row that
      // is not FRESH (spec 10.2). Surfaced as a block, not a soft warning.
      blockReasons.push({
        source: "scan",
        code: "FRESHNESS_POLICY_BLOCK",
        row_id: row.row_id,
        message:
          `freshness policy blocks ${row.row_id} (status ${row.status}); ` +
          `${defaultRequiredAction(row)}`
      });
      continue;
    }
    if (WARN_STATUSES.has(row.status)) {
      warnings.push(formatPrecommitWarning(row));
    }
  }

  const decision: PrecommitDecision =
    blockReasons.length > 0 ? "BLOCK" : warnings.length > 0 ? "WARN" : "OK";

  const messages: string[] = [
    ...blockReasons.map((reason) => `BLOCK: ${reason.message}`),
    ...warnings.map((warning) => warning.message)
  ];

  return {
    decision,
    exit_code: decision === "BLOCK" ? 1 : 0,
    block_reasons: blockReasons,
    warnings,
    messages
  };
}

// ---------------------------------------------------------------------------
// PR-summary formatter (spec 10.2 "textual summary"): a freshness status object
// -> a human-readable block of counts + each SUSPECT/INVALID row with its
// required action + the inferred Swift spans. Pure and deterministic.
// ---------------------------------------------------------------------------
export function formatFreshnessPrSummary(status: FreshnessStatus): string {
  const lines: string[] = [];
  lines.push(`USE-CASE FRESHNESS SUMMARY (policy: ${status.policy_mode})`);
  const summary = status.summary;
  lines.push(
    `fresh: ${summary.fresh}  suspect: ${summary.suspect}  unproven: ${summary.unproven}  ` +
      `unbound: ${summary.unbound}  invalid: ${summary.invalid}  policy_blocked: ${summary.policy_blocked}`
  );

  const invalidRows = status.rows.filter((row) => row.status === "INVALID");
  lines.push("");
  lines.push("INVALID rows:");
  if (invalidRows.length === 0) {
    lines.push("- (none)");
  } else {
    for (const row of invalidRows) {
      lines.push(`- ${row.row_id}  reason: ${firstReasonCode(row)}`);
      lines.push(`  required action: ${defaultRequiredAction(row)}`);
    }
  }

  const suspectRows = status.rows.filter(
    (row) => row.status === "SUSPECT" || row.status === "UNPROVEN" || row.status === "UNBOUND"
  );
  lines.push("");
  lines.push("SUSPECT / UNPROVEN / UNBOUND rows:");
  if (suspectRows.length === 0) {
    lines.push("- (none)");
  } else {
    for (const row of suspectRows) {
      lines.push(`- ${row.row_id}  status: ${row.status}  reason: ${firstReasonCode(row)}`);
      lines.push(`  required action: ${defaultRequiredAction(row)}`);
    }
  }

  // Inferred Swift spans (spec 8.2 / 10.2: CI must print inferred spans).
  const inferred = status.rows.flatMap((row) =>
    row.current_bindings
      .filter((binding) => binding.extent_kind === "swift_func_inferred")
      .map((binding) => ({ row_id: row.row_id, binding }))
  );
  if (inferred.length > 0) {
    for (const { row_id, binding } of inferred) {
      lines.push("");
      lines.push("INFERRED SWIFT SPAN");
      lines.push(`row: ${row_id}`);
      lines.push(`binding: ${binding.binding_slug}`);
      lines.push(`file: ${binding.file_path}`);
      lines.push(`span: lines ${binding.span_start_line}-${binding.span_end_line}`);
      lines.push(`span_sha256: ${binding.span_sha256}`);
    }
  }

  return lines.join("\n");
}

File: UseCasesCore/Sources/UseCasesCore/Markers/Commands/Precommit.swift
public enum PrecommitDecision: String, Equatable, Sendable {
  case block = "BLOCK"
  case warn = "WARN"
  case clear = "OK"
}

public enum PrecommitSource: String, Equatable, Sendable {
  case validateLedger = "validate-ledger"
  case scan
}

public struct PrecommitBlockReason: Equatable, Sendable {
  public let source: PrecommitSource
  public let code: String
  /// Present only for a reason tied to one row.
  public let rowIdentifier: String?
  public let message: String

  var jsonValue: JSONValue {
    var object = JSONObject([("source", .string(source.rawValue)), ("code", .string(code))])
    object["row_id"] = rowIdentifier.map(JSONValue.string)
    object["message"] = .string(message)
    return .object(object)
  }
}

public struct PrecommitWarning: Equatable, Sendable {
  public let rowIdentifier: String
  public let status: RowStatus
  public let reason: String
  public let requiredAction: String
  /// The loud, multi-line warning block (spec 10.1).
  public let message: String

  var jsonValue: JSONValue {
    .object(JSONObject([
      ("row_id", .string(rowIdentifier)),
      ("status", .string(status.rawValue)),
      ("reason", .string(reason)),
      ("required_action", .string(requiredAction)),
      ("message", .string(message)),
    ]))
  }
}

public struct PrecommitResult: Equatable, Sendable {
  public let decision: PrecommitDecision
  /// 1 when BLOCK, so the hook aborts; 0 otherwise.
  public let exitCode: Int
  public let blockReasons: [PrecommitBlockReason]
  public let warnings: [PrecommitWarning]
  /// Block reasons, then the loud warnings, ready to print.
  public let messages: [String]

  public var jsonValue: JSONValue {
    .object(JSONObject([
      ("decision", .string(decision.rawValue)),
      ("exit_code", .number(Double(exitCode))),
      ("block_reasons", .array(blockReasons.map(\.jsonValue))),
      ("warnings", .array(warnings.map(\.jsonValue))),
      ("messages", .array(messages.map(JSONValue.string))),
    ]))
  }
}

/// The slice of a validate-ledger result the precommit decision reads.
public struct PrecommitLedgerInput: Equatable, Sendable {
  public var isOK: Bool
  public var exitCode: Int
  public var evidenceValid: Bool
  public var registryValid: Bool
  public var appendOnly: Bool
  public var errors: [LedgerErrorReport]

  public init(
    isOK: Bool,
    exitCode: Int,
    evidenceValid: Bool,
    registryValid: Bool,
    appendOnly: Bool,
    errors: [LedgerErrorReport],
  ) {
    self.isOK = isOK
    self.exitCode = exitCode
    self.evidenceValid = evidenceValid
    self.registryValid = registryValid
    self.appendOnly = appendOnly
    self.errors = errors
  }

  public init(_ result: ValidateLedgerCommandResult) {
    self.init(
      isOK: result.isOK,
      exitCode: result.exitCode,
      evidenceValid: result.evidenceValid,
      registryValid: result.registryValid,
      appendOnly: result.appendOnly,
      errors: result.errors,
    )
  }
}

/// The slice of a scan result the precommit decision reads.
public struct PrecommitScanInput: Equatable, Sendable {
  public var exitCode: Int
  public var registryValid: Bool
  public var evidenceValid: Bool
  public var status: FreshnessStatus

  public init(
    exitCode: Int,
    registryValid: Bool,
    evidenceValid: Bool,
    status: FreshnessStatus,
  ) {
    self.exitCode = exitCode
    self.registryValid = registryValid
    self.evidenceValid = evidenceValid
    self.status = status
  }

  public init(_ result: ScanCommandResult) {
    self.init(
      exitCode: result.exitCode,
      registryValid: result.registryValid,
      evidenceValid: result.evidenceValid,
      status: result.status,
    )
  }
}

/// The precommit orchestrator and PR-summary formatter (precommit.ts, spec
/// 10.1 and 10.2). Pure: already-computed command results in, a decision or
/// text out.
public enum Precommit {
  /// `decidePrecommit`. BLOCK on a failed validate-ledger, a scan whose ledger
  /// or registry is invalid, a scan usage error (exit 2), an INVALID row or a
  /// policy-blocked row; WARN on a SUSPECT, UNPROVEN or UNBOUND row the policy
  /// lets through; otherwise OK.
  public static func decide(
    validateLedger: PrecommitLedgerInput,
    scan: PrecommitScanInput,
  ) -> PrecommitResult {
    var blockReasons = ledgerReasons(validateLedger)
    if !scan.registryValid || !scan.evidenceValid {
      blockReasons += ledgerIntegrityReasons(scan.status.integrityErrors)
    }
    if scan.exitCode == 2 {
      blockReasons.append(PrecommitBlockReason(
        source: .scan,
        code: "SCAN_USAGE_ERROR",
        rowIdentifier: nil,
        message: "scan failed with a usage/config/internal error (exit 2)",
      ))
    }

    var warnings: [PrecommitWarning] = []
    for row in scan.status.rows {
      if let reason = rowBlockReason(row) {
        blockReasons.append(reason)
      } else if [.suspect, .unproven, .unbound].contains(row.status) {
        warnings.append(warning(row))
      }
    }

    let decision: PrecommitDecision = !blockReasons.isEmpty
      ? .block
      : warnings.isEmpty ? .clear : .warn
    return PrecommitResult(
      decision: decision,
      exitCode: decision == .block ? 1 : 0,
      blockReasons: blockReasons,
      warnings: warnings,
      messages: blockReasons.map { reason in
        "BLOCK: \(reason.message)"
      } + warnings.map(\.message),
    )
  }

  /// Any validate-ledger failure blocks: one reason per error, or one generic
  /// reason when it reported none.
  private static func ledgerReasons(_ validateLedger: PrecommitLedgerInput)
    -> [PrecommitBlockReason]
  {
    guard !validateLedger.isOK else {
      return []
    }
    guard !validateLedger.errors.isEmpty else {
      return [PrecommitBlockReason(
        source: .validateLedger,
        code: "LEDGER_INVALID",
        rowIdentifier: nil,
        message: "validate-ledger reported the evidence ledger or binding registry as invalid",
      )]
    }
    return validateLedger.errors.map { error in
      PrecommitBlockReason(
        source: .validateLedger,
        code: error.code,
        rowIdentifier: nil,
        message: "validate-ledger (\(error.scope.rawValue)): \(error.message)",
      )
    }
  }

  /// An INVALID row blocks with its first reason; otherwise a policy-blocked
  /// row blocks as FRESHNESS_POLICY_BLOCK.
  private static func rowBlockReason(_ row: FreshnessRow) -> PrecommitBlockReason? {
    if row.status == .invalid {
      let code = firstReasonCode(row)
      return PrecommitBlockReason(
        source: .scan,
        code: code,
        rowIdentifier: row.rowIdentifier,
        message: "INVALID row \(row.rowIdentifier): \(code)",
      )
    }
    guard row.policyBlock else {
      return nil
    }
    return PrecommitBlockReason(
      source: .scan,
      code: "FRESHNESS_POLICY_BLOCK",
      rowIdentifier: row.rowIdentifier,
      message: "freshness policy blocks \(row.rowIdentifier) (status \(row.status.rawValue)); "
        + requiredAction(row),
    )
  }

  /// `formatPrecommitWarning`: the loud block spec 10.1 mandates.
  public static func warning(_ row: FreshnessRow) -> PrecommitWarning {
    let reason = firstReasonCode(row)
    let action = requiredAction(row)
    return PrecommitWarning(
      rowIdentifier: row.rowIdentifier,
      status: row.status,
      reason: reason,
      requiredAction: action,
      message: [
        "USE-CASE ROW \(row.status.rawValue)",
        "row: \(row.rowIdentifier)",
        "reason: \(reason)",
        "required action: \(action)",
      ].joined(separator: "\n"),
    )
  }

  /// `formatFreshnessPrSummary` (spec 10.2): the counts, each INVALID row, each
  /// SUSPECT / UNPROVEN / UNBOUND row with its required action, and every
  /// inferred Swift span.
  public static func pullRequestSummary(_ status: FreshnessStatus) -> String {
    let summary = status.summary
    var lines = [
      "USE-CASE FRESHNESS SUMMARY (policy: \(status.policyMode.rawValue))",
      "fresh: \(summary.fresh)  suspect: \(summary.suspect)  unproven: \(summary.unproven)  "
        + "unbound: \(summary.unbound)  invalid: \(summary.invalid)  "
        + "policy_blocked: \(summary.policyBlocked)",
    ]

    let invalidRows = status.rows.filter { row in
      row.status == .invalid
    }
    lines += ["", "INVALID rows:"]
    if invalidRows.isEmpty {
      lines.append("- (none)")
    }
    for row in invalidRows {
      lines.append("- \(row.rowIdentifier)  reason: \(firstReasonCode(row))")
      lines.append("  required action: \(requiredAction(row))")
    }

    let attentionRows = status.rows.filter { row in
      [.suspect, .unproven, .unbound].contains(row.status)
    }
    lines += ["", "SUSPECT / UNPROVEN / UNBOUND rows:"]
    if attentionRows.isEmpty {
      lines.append("- (none)")
    }
    for row in attentionRows {
      lines.append(
        "- \(row.rowIdentifier)  status: \(row.status.rawValue)  reason: \(firstReasonCode(row))",
      )
      lines.append("  required action: \(requiredAction(row))")
    }

    for row in status.rows {
      for binding in row.currentBindings where binding.extentKind == .swiftFunctionInferred {
        lines += [
          "",
          "INFERRED SWIFT SPAN",
          "row: \(row.rowIdentifier)",
          "binding: \(binding.bindingSlug)",
          "file: \(binding.filePath)",
          "span: lines \(binding.span.startLine)-\(binding.span.endLine)",
          "span_sha256: \(binding.span.sha256)",
        ]
      }
    }
    return lines.joined(separator: "\n")
  }

  /// Ledger-level integrity errors — those carrying no `row_id` member at all
  /// (an explicit null is a member) — or one generic reason when there are
  /// none.
  private static func ledgerIntegrityReasons(_ errors: [JSONObject]) -> [PrecommitBlockReason] {
    let ledgerLevel = errors.filter { error in
      error["row_id"] == nil
    }
    guard !ledgerLevel.isEmpty else {
      return [PrecommitBlockReason(
        source: .scan,
        code: "LEDGER_INVALID",
        rowIdentifier: nil,
        message: "scan reported the evidence ledger or binding registry as invalid",
      )]
    }
    return ledgerLevel.map { error in
      // Freshness writes every integrity code and message as a string; a null
      // or absent message falls back to the code (`??`), an empty one does not.
      let code = error["code"]?.stringValue ?? ""
      let message = error["message"]?.stringValue ?? code
      return PrecommitBlockReason(
        source: .scan,
        code: code,
        rowIdentifier: nil,
        message: "scan ledger/registry: \(message)",
      )
    }
  }

  /// `row.reasons[0]?.code ?? row.status`. Freshness writes every reason code
  /// as a string.
  private static func firstReasonCode(_ row: FreshnessRow) -> String {
    row.reasons.first?["code"]?.stringValue ?? row.status.rawValue
  }

  /// `defaultRequiredAction`: the row's own action when it is a non-empty
  /// string (truthiness), else `uc bind` for UNBOUND and `uc prove` otherwise.
  private static func requiredAction(_ row: FreshnessRow) -> String {
    if let action = row.requiredAction, !action.isEmpty {
      return action
    }
    return row.status == .unbound
      ? "uc bind --row \(row.rowIdentifier)"
      : "uc prove --row \(row.rowIdentifier)"
  }
}
